docs/source/notes/numpy_compatibility.rst

```diff
@@ -0,0 +1,221 @@
+fundamentally different types. However, in the interests of being accessible to
+the wider scientific python ecosystem, there are implicit and explicit
+conversion methods. The major considerations to keep in mind are:
+
+- **Where** is the structure allocated?
+    * This may be on a compute device (e.g. a GPU) or the CPU
+- **What** is the underlying data?
+    * In particular, several functions will either return a copy or a view
+- **How** are modifications to the data handled?
+    * The data returned may be read-only, or may be writable
+
+From the end-user perspective, it is most desirable to have zero copy methods,
+however, for exploratory analyses and their visualizations in particular, it is
+often more important to have guaranteed conversions, even if copies are created.
+
+Basic interoperability
+------------------------
+
+NumPy arrays have no conception of the GPU, and so all conversions require a
+``cpu`` tensor. Similarly, the **computational graph** cannot be represented in
+```numpy``, and so the pre-requisite to have an operation or object without one,
+i.e. only leaf nodes are supported.
+
+.. code-block:: python
+
+   import torch
+   b_torch = torch.ones(3)
+   assert(b_torch.requires_grad == False)
+   assert(b_torch.is_cuda == False)
+   b_torch.numpy() # Fine, will work
+
+The interoperability guidelines places a strong emphasis on correctness over
+convenience. As the conversion to a NumPy array may potentially be a lossy
+operation (the computational graph), there are no inbuilt operations to do
+something which unconditionally returns a NumPy array.
+
+.. code-block:: python
+
+   # Never do this
+   def uncond_numpy(torch_tensor):
+        if torch_tensor.is_cuda:
+           warnings.warn("Moving back to CPU")
+           torch_tensor = torch_tensor.cpu()
+        if torch_tensor.requires_grad == False:
+            return torch_tensor.numpy()
+        else:
+            warnings.warn("Calling detach, probably lossy!")
+            return torch_tensor.detach().numpy()
+
+Operations
+------------
+
+All ``torch`` operators will helpfully fail with a ``TypeError`` if called with
+``numpy`` arrays. However, for **numpy methods**, using a ``torch.Tensor`` with
+an ``np.ndarray`` will return a ``torch.Tensor`` without creating any additional
+copies.
+
+ - Due to the :meth:`torch.Tensor.__array__()` implementation, a
+   ``np.ndarray`` which shares memory with the ``torch.Tensor`` is used for the
+   operation.
+ - The return type functionality is defined by
+   :meth:`torch.Tensor.__array_wrap__()`, and calls ``torch.from_numpy()``
+   internally.
+
+.. code-block:: python
+
+   import torch
+   import numpy as np
+   a_np = np.ones(3) # dtype=float64
+   b_torch = torch.ones(3) # dtype=float32
+   torch.add(a_np, b_torch) # TypeError
+   b_torch + a_np # TypeError
+   a_np + b_torch # OK but awkward
+   c_tmp = np.add(a_np, b_torch) # OK
+   # c_tmp.dtype is torch.float64
```

Comment:
To give a concrete example of the doctest format:
```python
>>> c = np.add(a_np, b_torch)
>>> c.dtype
torch.float64
```
Much more readable than these comments.

Reply:
_You want to show the example, and then explain why that behaves as it does I think. Here is what I would write:_

```python
>>> x + t
...
TypeError: Concatenation operation is not implemented for NumPy arrays, use np.concatenate() instead. Please do not rely on this error; it may not be given on all Python implementations.

>>> t + x
tensor([4., 4., 4.], dtype=torch.float64)
```
The reason that `x + t` errors here is that, due to the rules of the Python language for operator support, PyTorch is called here. To be exact, `Tensor.__add__` is called - and that does not accept NumPy arrays.

`t + x` on the other hand calls `ndarray.__add__`. NumPy will then convert the tensor `t` to a `numpy.ndarray` (a zero-copy operation), calculate the result for adding two arrays, and then convert the output back to a `torch.Tensor` (this last step happens because `Tensor` implements an `__array_wrap__` method, instructing NumPy to do that conversion).